File: Frameworks/flutter-local/test/models/ods_app_setting_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:ods_flutter_local/models/ods_app_setting.dart';

void main() {
  // =========================================================================
  // fromJson
  // =========================================================================

  group('OdsAppSetting.fromJson', () {
    test('full valid setting with options as list', () {
      final setting = OdsAppSetting.fromJson({
        'label': 'Language',
        'type': 'select',
        'default': 'en',
        'options': ['en', 'fr', 'de'],
      });
      expect(setting.label, 'Language');
      expect(setting.type, 'select');
      expect(setting.defaultValue, 'en');
      expect(setting.options, ['en', 'fr', 'de']);
    });

    test('default label is empty string when null', () {
      final setting = OdsAppSetting.fromJson({
        'type': 'text',
        'default': 'hello',
      });
      expect(setting.label, '');
    });

    test('default type is text when null', () {
      final setting = OdsAppSetting.fromJson({
        'label': 'Name',
        'default': 'World',
      });
      expect(setting.type, 'text');
    });

    test('default defaultValue is empty string when null', () {
      final setting = OdsAppSetting.fromJson({
        'label': 'Name',
        'type': 'text',
      });
      expect(setting.defaultValue, '');
    });

    test('options as comma-separated string are split and trimmed', () {
      final setting = OdsAppSetting.fromJson({
        'label': 'Size',
        'type': 'select',
        'default': 'M',
        'options': 'S, M , L, XL',
      });
      expect(setting.options, ['S', 'M', 'L', 'XL']);
    });

    test('options missing returns null', () {
      final setting = OdsAppSetting.fromJson({
        'label': 'Username',
        'type': 'text',
        'default': '',
      });
      expect(setting.options, isNull);
    });
  });
}

File: Frameworks/react-web/tests/unit/models/ods-app-setting.test.ts
import { describe, it, expect } from 'vitest'
import { parseAppSetting } from '../../../src/models/ods-app-setting.ts'

// ===========================================================================
// OdsAppSetting model tests
// ===========================================================================

describe('parseAppSetting', () => {
  // -------------------------------------------------------------------------
  // Full valid setting
  // -------------------------------------------------------------------------

  it('parses a full valid setting', () => {
    const setting = parseAppSetting({
      label: 'Language',
      type: 'select',
      default: 'en',
      options: ['en', 'fr', 'de'],
    })
    expect(setting.label).toBe('Language')
    expect(setting.type).toBe('select')
    expect(setting.defaultValue).toBe('en')
    expect(setting.options).toEqual(['en', 'fr', 'de'])
  })

  // -------------------------------------------------------------------------
  // Default values
  // -------------------------------------------------------------------------

  it('default label is empty string', () => {
    const setting = parseAppSetting({})
    expect(setting.label).toBe('')
  })

  it('default type is text', () => {
    const setting = parseAppSetting({})
    expect(setting.type).toBe('text')
  })

  it('default defaultValue is empty string', () => {
    const setting = parseAppSetting({})
    expect(setting.defaultValue).toBe('')
  })

  // -------------------------------------------------------------------------
  // Options handling
  // -------------------------------------------------------------------------

  it('options as array', () => {
    const setting = parseAppSetting({ options: ['a', 'b', 'c'] })
    expect(setting.options).toEqual(['a', 'b', 'c'])
  })

  it('options as comma-separated string', () => {
    const setting = parseAppSetting({ options: 'red, green, blue' })
    expect(setting.options).toEqual(['red', 'green', 'blue'])
  })

  it('options missing returns undefined', () => {
    const setting = parseAppSetting({ label: 'Name' })
    expect(setting.options).toBeUndefined()
  })
})
